ls1intum/Artemis
# `Communication`: Improve performance of communication overview page with virtual scrolling

### Checklist
#### General



#### Client


### Motivation and Context
Virtual scrolling is the way of displaying a set of elements by only having a couple of the items that can fit into the list container at a time, aiming to improve the overall performance by passing a minimal amount of items to the DOM tree. As users scroll through the items on their screen, the subset of these items available in the DOM tree is manipulated so that they still have a smooth scrolling experience.

https://material.angular.io/cdk/scrolling/overview#virtual-scrolling

### Description
The solution provided by Angular is limited to items whose heights are of fixed size. Therefore, we have to come up with our own implementation. In this PR, we introduce the new VirtualScroll component. For all rendered posts, the heights are cached to calculate the currently rendered item. 

When posts are rendered after being removed from the DOM tree, their heights shrink if they previously had collapsible parts open. In our implementation for all posts to be removed from the DOM tree, we calculate the collapsible heights first and update the current scrolling distance from the top of the page so that users have a smooth scrolling experience.

### Steps for Testing


Note: Please don't test on TS1 Canberk's course. Something seems to be broken there. If you need a course with lots of posts, you may use TS2 Canberk's course as an alternative.

Prerequisites:
- 1 Instructor
- 1 Course with discussions enabled

1. Log in to Artemis
2. Navigate to Course Communication overview page
3. Make sure posts scroll as expected
4. React to posts, answer them, and edit them. Make sure that the behavior is as expected.
5. Try out different sorting and filtering combinations, and make sure that the page behaves as expected.

### Review Progress






#### Code Review
#### Manual Tests


### Test Coverage


| Class/File | Branch | Line |
|------------|-------:|-----:|
| course-discussion.component.ts | 81% | 95% |
| virtual-scroll-render-event.class.ts | 100% | 100% |
| virtual-scroll.component.ts  | 69% | 85% |


### Screenshots
[Virtual Scroll.webm](https://user-images.githubusercontent.com/67344158/202851986-0f34618a-a29b-4e31-a0b0-ca2d7b9fdb12.webm)
